extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java

```diff
@@ -298,4 +336,119 @@ private ResultHandle generateSort(Integer sortParameterIndex, MethodCreator meth
+        DotName interfaceName = interfaceInfo.name();
+        DotName customImplName = customReturnTypes.get(interfaceName);
+
+        Map<String, FieldDescriptor> fields = new HashMap<>(fieldNames.size());
+        if (customImplName == null) {
+            customImplName = GenerationUtil.createImplDotName(interfaceName);
+            try (ClassCreator implClassCreator = ClassCreator.builder().classOutput(nonBeansClassOutput)
+                    .interfaces(interfaceName.toString()).className(customImplName.toString())
+                    .build()) {
+
+                List<MethodInfo> methods = interfaceInfo.methods();
+
+                // Work with methods in declared order
+                for (int i = 0; i < methods.size(); i++) {
+                    MethodInfo method = methods.get(i);
+
+                    if (!method.parameters().isEmpty()) {
+                        throw new IllegalArgumentException("Method " + method.name() + " of interface " + interfaceName
+                                + " is not a getter method since it defined parameters");
+                    }
+                    Type returnType = method.returnType();
+                    if (returnType.kind() == Type.Kind.VOID) {
+                        throw new IllegalArgumentException("Method " + method.name() + " of interface " + interfaceName
+                                + " is not a getter method since it returns void");
+                    }
+
+                    String getterName = method.name();
+                    String propertyName = JavaBeanUtil.getPropertyNameFromGetter(getterName);
+                    DotName returnTypeName = method.returnType().name();
+                    DotName fieldTypeName = GenerationUtil.getPrimitiveTypeName(returnTypeName);
+
+                    if (!fieldNames.contains(propertyName)) {
+                        throw new IllegalArgumentException("Method " + method.name() + " of interface " + interfaceName
+                                + " is not a getter method for a field defined in the original query " + fieldNames);
+                    }
+
+                    FieldDescriptor field = implClassCreator.getFieldCreator(propertyName, fieldTypeName.toString())
+                            .getFieldDescriptor();
+                    fields.put(propertyName, field);
+
+                    // create getter (based on the interface)
+                    try (MethodCreator getter = implClassCreator.getMethodCreator(getterName, returnType.toString())) {
+                        getter.setModifiers(Modifier.PUBLIC);
+                        getter.returnValue(getter.readInstanceField(field, getter.getThis()));
+                    }
+                }
+
+                // Generate a static 'convert' method, to convert Object[] to field values
+                try (MethodCreator convert = implClassCreator.getMethodCreator("convert",
```

Comment:
Although this does do what it's intended to do, I suspect however that it will be problematic when the same Class is returned from multiple repository methods that don't use the same ordering for fields.

In you test you have:

```
@Query(value = "SELECT publicationYear as publicationYear, COUNT(*) as count FROM Book GROUP BY publicationYear")
    List<BookCountByYear> findAllByPublicationYear();
```

What happens if I do this:

```
@Query(value = "SELECT COUNT(*) as count, publicationYear as publicationYear FROM Book GROUP BY publicationYear")
    List<BookCountByYear> findAllByPublicationYear2();
```

I actually did try this and it failed :(

Reply:
What I think you'll need to do is take into account the `fieldNames` in the `convert` method by adding it as a parameter and then doing some magic to work out the proper order

Reply:
I think I'd have to mark the query uniquely, and include that in a separate convert method. The order of the columns is significant, so while those have the same custom return type, it is probably safer to treat as two separate implementations. .. maybe. I will have to see if I still have the field order available when calling convert later.